File: application/src/main/java/de/pbauerochse/worklogviewer/WorklogViewer.java
package de.pbauerochse.worklogviewer;

import de.pbauerochse.worklogviewer.fx.tasks.TaskRunner;
import de.pbauerochse.worklogviewer.fx.theme.ThemeChangeListener;
import de.pbauerochse.worklogviewer.settings.Settings;
import de.pbauerochse.worklogviewer.settings.SettingsUtil;
import de.pbauerochse.worklogviewer.settings.SettingsViewModel;
import de.pbauerochse.worklogviewer.util.ExceptionUtil;
import de.pbauerochse.worklogviewer.util.FormattingUtil;
import javafx.application.Application;
import javafx.fxml.FXMLLoader;
import javafx.scene.Parent;
import javafx.scene.Scene;
import javafx.stage.Stage;
import javafx.stage.WindowEvent;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.nio.charset.Charset;
import java.nio.charset.StandardCharsets;
import java.util.Locale;
import java.util.TimeZone;

/**
 * Main class to start the Java FX UI
 */
public class WorklogViewer extends Application {

    private static final Logger LOGGER = LoggerFactory.getLogger(WorklogViewer.class);

    private static WorklogViewer instance;

    public static WorklogViewer getInstance() {
        if (instance == null) {
            throw ExceptionUtil.getIllegalStateException("exceptions.main.instance");
        }
        return instance;
    }

    public static void main(String[] args) {
        launch(args);
    }

    private Stage primaryStage;

    @Override
    public void stop() {
        SettingsUtil.saveSettings();
        TaskRunner.getEXECUTOR().shutdownNow();
    }

    @Override
    public void start(Stage primaryStage) throws Exception {
        instance = this;
        this.primaryStage = primaryStage;

        Settings settings = SettingsUtil.getSettings();
        SettingsViewModel settingsViewModel = SettingsUtil.getSettingsViewModel();

        LOGGER.info("Java Version: {}", System.getProperty("java.version"));
        LOGGER.info("Default Locale: {}", Locale.getDefault());
        LOGGER.info("Default Charset: {}", Charset.defaultCharset());
        LOGGER.info("Default TimeZone: {}", TimeZone.getDefault().toZoneId());
        LOGGER.info("Theme: {}", settings.getTheme());

        FXMLLoader loader = new FXMLLoader(StandardCharsets.UTF_8);
        loader.setResources(FormattingUtil.RESOURCE_BUNDLE);

        Parent root = loader.load(WorklogViewer.class.getResource("/fx/views/main.fxml").openStream());
        Scene mainScene = new Scene(root, settings.getWindowSettings().getWidth(), settings.getWindowSettings().getHeight());

        mainScene.getStylesheets().add("/fx/css/base-styling.css");
        mainScene.getStylesheets().add(settingsViewModel.getTheme().getStylesheet());
        settingsViewModel.themeProperty().addListener(new ThemeChangeListener(mainScene));

        primaryStage.setTitle("YouTrack Worklog Viewer " + FormattingUtil.getFormatted("release.version"));
        primaryStage.setScene(mainScene);
        primaryStage.setX(settings.getWindowSettings().getPositionX());
        primaryStage.setY(settings.getWindowSettings().getPositionY());
        primaryStage.setWidth(settings.getWindowSettings().getWidth());
        primaryStage.setHeight(settings.getWindowSettings().getHeight());
        primaryStage.show();

        primaryStage.widthProperty().addListener((observable, oldValue, newValue) -> settings.getWindowSettings().setWidth(newValue.intValue()));
        primaryStage.heightProperty().addListener((observable, oldValue, newValue) -> settings.getWindowSettings().setHeight(newValue.intValue()));
        primaryStage.xProperty().addListener((observable, oldValue, newValue) -> settings.getWindowSettings().setPositionX(newValue.intValue()));
        primaryStage.yProperty().addListener((observable, oldValue, newValue) -> settings.getWindowSettings().setPositionY(newValue.intValue()));
    }

    public void requestShutdown() {
        LOGGER.debug("Shutdown requested");
        primaryStage.fireEvent(new WindowEvent(primaryStage, WindowEvent.WINDOW_CLOSE_REQUEST));
    }

}

File: application/src/main/java/de/pbauerochse/worklogviewer/WorklogViewer.kt
package de.pbauerochse.worklogviewer

import de.pbauerochse.worklogviewer.fx.theme.ThemeChangeListener
import de.pbauerochse.worklogviewer.preloader.Preloader
import de.pbauerochse.worklogviewer.settings.SettingsUtil.saveSettings
import de.pbauerochse.worklogviewer.settings.SettingsUtil.settings
import de.pbauerochse.worklogviewer.settings.SettingsUtil.settingsViewModel
import de.pbauerochse.worklogviewer.tasks.Tasks.shutdown
import de.pbauerochse.worklogviewer.util.FormattingUtil
import javafx.application.Application
import javafx.application.Platform
import javafx.beans.value.ObservableValue
import javafx.fxml.FXMLLoader
import javafx.scene.Parent
import javafx.scene.Scene
import javafx.scene.image.Image
import javafx.stage.Stage
import javafx.stage.WindowEvent
import org.slf4j.LoggerFactory
import java.awt.Taskbar
import java.awt.Toolkit
import java.io.InputStream
import java.nio.charset.Charset
import java.nio.charset.StandardCharsets
import java.util.*

/**
 * Main class to start the Java FX UI
 */
class WorklogViewer : Application() {

    private lateinit var primaryStage: Stage

    @Throws(Exception::class)
    override fun start(stage: Stage) {
        instance = this

        val settings = settings
        val settingsViewModel = settingsViewModel

        LOGGER.info("Java Version: {}", System.getProperty("java.version"))
        LOGGER.info("Default Locale: {}", Locale.getDefault())
        LOGGER.info("Default Charset: {}", Charset.defaultCharset())
        LOGGER.info("Default TimeZone: {}", TimeZone.getDefault().toZoneId())
        LOGGER.info("Theme: {}", settings.theme)

        Preloader.preload()

        val loader = FXMLLoader(StandardCharsets.UTF_8).apply {
            resources = FormattingUtil.RESOURCE_BUNDLE
        }

        val root = classPathResource("/fx/views/main.fxml")!!.use {
            loader.load<Parent>(it)
        }

        val mainScene = Scene(root, settings.windowSettings.width.toDouble(), settings.windowSettings.height.toDouble()).apply {
            stylesheets.add("/fx/css/base-styling.css")
            stylesheets.add(settingsViewModel.themeProperty.get().stylesheet)
        }
        settingsViewModel.themeProperty.addListener(ThemeChangeListener(mainScene))

        primaryStage = stage.apply {
            title = "YouTrack Worklog Viewer ${FormattingUtil.getFormatted("release.version")}"
            icons.addAll(listOf(
                classPathResource("/fx/img/icons/logo-128.png")!!.use { Image(it) },
                classPathResource("/fx/img/icons/logo-64.png")!!.use { Image(it) },
                classPathResource("/fx/img/icons/logo-32.png")!!.use { Image(it) }
            ))
            scene = mainScene

            x = settings.windowSettings.positionX.toDouble()
            y = settings.windowSettings.positionY.toDouble()
            width = settings.windowSettings.width.toDouble()
            height = settings.windowSettings.height.toDouble()
        }
        primaryStage.show()
        setDockAndTaskbarIcon()

        // settings listener
        stage.widthProperty().addListener { _: ObservableValue<out Number>?, _: Number?, newValue: Number -> settings.windowSettings.width = newValue.toInt() }
        stage.heightProperty().addListener { _: ObservableValue<out Number>?, _: Number?, newValue: Number -> settings.windowSettings.height = newValue.toInt() }
        stage.xProperty().addListener { _: ObservableValue<out Number>?, _: Number?, newValue: Number -> settings.windowSettings.positionX = newValue.toInt() }
        stage.yProperty().addListener { _: ObservableValue<out Number>?, _: Number?, newValue: Number -> settings.windowSettings.positionY = newValue.toInt() }
    }

    /**
     * MacOS does not show the icon in the dock
     *
     * Causes a "Gdk-WARNING **: 12:55:57.537: XSetErrorHandler() called with a GDK error trap pushed. Don't do that."
     * Warning on GTK3...don't know how to get rid of it, yet
     */
    private fun setDockAndTaskbarIcon() {
        Platform.runLater {
            if (Taskbar.isTaskbarSupported()) {
                try {
                    val icon = Toolkit.getDefaultToolkit().getImage(WorklogViewer::class.java.getResource("/fx/img/icons/logo-128.png"))
                    Taskbar.getTaskbar()?.iconImage = icon
                } catch (e: Exception) {
                    LOGGER.warn("Could not set Taskbar image", e)
                }
            }
        }
    }

    override fun stop() {
        saveSettings()
        shutdown()
    }

    fun requestShutdown() {
        LOGGER.debug("Shutdown requested")
        primaryStage.fireEvent(WindowEvent(primaryStage, WindowEvent.WINDOW_CLOSE_REQUEST))
    }

    private fun classPathResource(resource: String): InputStream? = WorklogViewer::class.java.getResource(resource)?.openStream()

    companion object {
        private val LOGGER = LoggerFactory.getLogger(WorklogViewer::class.java)
        lateinit var instance: WorklogViewer
    }
}